File: examples/language-features/CSharp/dotnet/7.0/10.0/library/CSharp3/AnonymousTypes/AnonymousTypes.cs
using System.Collections.Generic;

namespace Net7_CSharp10_Library.CSharp3.AnonymousTypes
{
    public class AnonymousTypeSamples
    {
        // An anonymous type has no writable name, so var is mandatory.
        public static string Describe()
        {
            var point = new { X = 1, Y = 2 };
            return point.X + "," + point.Y;
        }

        // Two anonymous types sharing property names, types, and order compile
        // to one generated type, and the generated Equals compares by value.
        public static bool StructurallyEqual()
        {
            var first = new { Name = "a", Count = 1 };
            var second = new { Name = "a", Count = 1 };
            return first.Equals(second);
        }

        // The properties are read-only; projection is the usual reason to reach
        // for one.
        public static List<string> ProjectNames(string[] names)
        {
            var projected = new List<string>();
            foreach (string name in names)
            {
                var item = new { Original = name, Length = name.Length };
                projected.Add(item.Original + ":" + item.Length);
            }

            return projected;
        }
    }
}

File: examples/language-features/VB.NET/dotNetFramework/v4.8/src/Baseline/AnonymousTypes/AnonymousTypes.vb
Option Strict On
Option Infer On

Imports System.Collections.Generic

Namespace Baseline.AnonymousTypes
    Public Module Anonymous
        ' An anonymous type has no writable name, so Option Infer must be On.
        Public Function Describe() As String
            Dim point = New With {.X = 1, .Y = 2}
            Return point.X.ToString() & "," & point.Y.ToString()
        End Function

        ' VB anonymous-type members are MUTABLE by default, unlike C#'s, which
        ' are always read-only. Key marks a member immutable and includes it in
        ' the generated Equals and GetHashCode.
        Public Function MutableByDefault() As Integer
            Dim item = New With {.Count = 1}
            item.Count = 2
            Return item.Count
        End Function

        ' Only Key members take part in equality, so these two are equal even
        ' though their Count differs.
        Public Function KeyEquality() As Boolean
            Dim first = New With {Key .Name = "a", .Count = 1}
            Dim second = New With {Key .Name = "a", .Count = 2}
            Return first.Equals(second)
        End Function

        Public Function Projection(names As String()) As List(Of String)
            Dim projected As New List(Of String)()

            For Each name As String In names
                Dim item = New With {Key .Original = name, .Length = name.Length}
                projected.Add(item.Original & ":" & item.Length.ToString())
            Next

            Return projected
        End Function
    End Module
End Namespace
